CRM/Aoservicelisting/Form/ProviderApplication.php

```diff
@@ -34,8 +34,33 @@ class CRM_Aoservicelisting_Form_ProviderApplication extends CRM_Core_Form {
   public function preProcess() {
     parent::preProcess();
     CRM_Core_Resources::singleton()->addStyleFile('biz.jmaconsulting.aoservicelisting', 'css/providerformstyle.css');
-    // for testing purpose consider cid value from url
-    $loggedInContactId = CRM_Utils_Request::retrieve('cid', 'Positive', $this, FALSE) ?: $this->getLoggedInUserContactID();
+    $invalidUser = FALSE;
+    $loggedInContactId = $this->getLoggedInUserContactID();
+    $cid = CRM_Utils_Request::retrieve('cid', 'Positive', $this, FALSE);
+    $cs = CRM_Utils_Request::retrieve('cs', 'String', $this, FALSE);
+    if (empty($cs) && !empty($_GET['cs'])) {
+      $cs = $_GET['cs'];
+    }
+    if (empty($cid) && !empty($_GET['cid'])) {
+      $cid = $_GET['cid'];
+    }
+    // Validate checksum only if specified in the URL, for non-logged in users.
+    if (!empty($cid) && !empty($cs)) {
+      if (!CRM_Contact_BAO_Contact_Utils::validChecksum($cid, $cs)) {
+        $invalidUser = TRUE;
+      }
+      else {
+        // This is a valid user.
+        $loggedInContactId = $cid;
+      }
+    }
```

Comment:
@Edzelopez if the $cid is not provided in link then we should consider the loggedIncontactId right? 

like here it should be 
```
else {
$cid = $loggedInContactId;
}
```
right ?

Reply:
Not quite. If the $cid is not provided in the link, the service listing application should be treated as a create form. On L38 above, we are already checking the fact that the user is logged in, and so a logged in user will have no reason to have $cid in the link.
If we add the else part, L58 below will consider the form to be submitted by an invalid user.

Reply:
agree.